### PR TITLE
fix(executor): Handle sidecar killing in a process-namespace-shared pod

### PR DESCRIPTION
**Pull request by dtaniwaki**
Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

I found sending `SIGTERM` fails to stop a container if `shareProcessNamespace: true` is set with the k8sapi executor because the PID of a container isn't `1`. I understand it doesn't make sense using the option without the pns executor, but my company's cluster forcibly set `shareProcessNamespace: true` but we haven't evaluated the pns executor yet. We should at least report the invalid option to users.

I also improved error messages during sidecar killing.

### CODE REVIEW

File: workflow/executor/common/common.go [hunk 1]
```diff
@@ -80,6 +80,9 @@ func TerminatePodWithContainerID(c KubernetesClientInterface, containerID string
 		log.Infof("Container %s is already terminated: %v", container.ContainerID, container.State.Terminated.String())
 		return nil
 	}
+	if pod.Spec.ShareProcessNamespace != nil && *pod.Spec.ShareProcessNamespace {
+		return fmt.Errorf("cannot terminate a process-namespace-shared Pod %s", pod.Name)
```
Comment (dtaniwaki) [inline workflow/executor/common/common.go:84]: If we return immediately with an error during waiting sidecars here, the wait container exit with `0` and the pod succeeds. It's also the same in other error returns in this method too.

Comment (alexec) [reply workflow/executor/common/common.go:84]: test?

Comment (dtaniwaki) [reply workflow/executor/common/common.go:84]: Added a unit test of this method. I think we should have e2e tests for all the executors. If we had, [this functional test](https://github.com/argoproj/argo/blob/master/test/e2e/functional/sidecar-force-kill.yaml) found the problem.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
LGTM

File: workflow/executor/common/common.go (hunk 1 above)
Comment (alexec) [reply workflow/executor/common/common.go:84]: We have e2e smoke test for 3 of the executors?

Comment (dtaniwaki) [reply workflow/executor/common/common.go:84]: Ah, you're right. The problem occurs on the combination of k8sapi and `shareProcessNamespace`.

**Comment by alexec** [comment]
Merged into https://github.com/argoproj/argo/releases/tag/v2.12.0-rc4

**Pull request merged**